src/sources.cpp

```diff
@@ -436,4 +436,102 @@ void fields::add_volume_source(component c, const src_time &src, const volume &w
   require_component(c);
 }
 
+/* Given the field-strength vector E0 for a Gaussian beam centered at x0 with beam waist w0 and propagation direction kdir,
+   compute the E/H fields EH[6] (6 components) at x for a frequency freq in the homogeneous 3d medium eps and mu.
+
+   Adapted from code by M. T. Homer Reid in his SCUFF-EM package
+   (file scuff-em/src/libs/libIncField/GaussianBeam.cc), which is GPL v2+. */
+
+void fields::gaussianbeam_3d(std::complex<double> *EH, const vec &x0, const vec &x,
```

Comment:
Probably this doesn't need to be a method of the `fields` class since it doesn't use the fields.